websockets/handlers/msg_channel_wsh.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/python
+import json
+import logging
+import urllib
+import threading
+import traceback
+from queue import Empty
+
+from mod_pywebsocket import stream, msgutil
+from wptserve import stash as stashmod
+
+logger = logging.getLogger()
+
+address, authkey = stashmod.load_env_config()
+stash = stashmod.Stash("msg_channel", address=address, authkey=authkey)
+
+def log(uuid, msg):
+    logger.debug(f"{uuid}: {msg}")
+
+
+def web_socket_do_extra_handshake(request):
+    return
+
+
+def web_socket_transfer_data(request):
+    uuid, direction = parse_request(request)
+    log(uuid, f"Got web_socket_transfer_data {direction}")
+
+    with stash.lock:
+        value = stash.take(uuid)
+        if value is None:
+            queue = stash.get_queue()
+            if direction == "read":
+                has_reader = True
+                writer_count = 0
+            else:
+                has_reader = False
+                writer_count = 1
+        else:
+            queue, has_reader, writer_count = value
+            if direction == "read":
+                if has_reader:
+                    raise ValueError("Tried to start multiple readers for the same queue")
+            else:
+                writer_count += 1
+
+        stash.put(uuid, (queue, has_reader, writer_count))
+
+    if direction == "read":
+        run_read(request, uuid, queue)
+    elif direction == "write":
+        run_write(request, uuid, queue)
+
+    log(uuid, f"transfer_data loop exited {direction}")
+    close_channel(uuid, direction)
+
+
+def web_socket_passive_closing_handshake(request):
+    uuid, direction = parse_request(request)
+    log(uuid, f"Got web_socket_passive_closing_handshake {direction}")
+
+    if direction == "read":
+        with stash.lock:
+            data = stash.take(uuid)
+            stash.put(uuid, data)
+        if data is not None:
+            queue = data[0]
+            queue.put(("close", None))
+
+    return request.ws_close_code, request.ws_close_reason
+
+
+def parse_request(request):
+    query = request.unparsed_uri.split('?')[1]
+    GET = dict(urllib.parse.parse_qsl(query))
+    uuid = GET["uuid"]
+    direction = GET["direction"]
+    return uuid, direction
+
+
+def wait_for_close(request, uuid, queue):
+    closed = False
+    while not closed:
+        try:
+            line = request.ws_stream.receive_message()
```

Comment:
Is it really a line?

```suggestion
            msg = request.ws_stream.receive_message()
```